### PR TITLE
Idp 995 add only non empty guids and strings attributes

### PR DESCRIPTION
**Pull request by nicholas-renaud**
## Description of changes
Adding the ContainsOnlyNonEmptyGuidsAttribute and ContainsOnlyNonEmptyStringsAttribute to cover some use cases that come up in the authentication service and need manual intervention to validate.

## Additional checks
Added a line of missing documentation and reformatted the description grid to accomodate a larger attribute name

- [ x ] Updated the documentation of the project to reflect the changes
- [ x ] Added new tests that cover the code changes

### CODE REVIEW

File: src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyGuidsAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Workleap.ComponentModel.DataAnnotations;
+
+[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Parameter)]
+public sealed class ContainsOnlyNonEmptyGuidsAttribute : ValidationAttribute
+{
+    internal const string ErrorMessageFormat = "The field {0} must be an collection that contains only non-empty GUIDs";
```
Comment (PrincessMadMath) [inline src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyGuidsAttribute.cs:11]:
```suggestion
    internal const string ErrorMessageFormat = "The field {0} must be a collection that contains only non-empty GUIDs";
```

File: src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyStringsAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    internal const string ErrorMessageFormat = "The field {0} must be a collection that contains only non-empty Strings";
+    
+    public ContainsOnlyNonEmptyStringsAttribute() : base(ErrorMessageFormat)
+    {
+    }
+
+    public override bool IsValid(object? value) => value switch
+    {
+        null => true,
+        IEnumerable<string> enumerable => enumerable.All(x => !string.IsNullOrWhiteSpace(x)),
```
Comment (PrincessMadMath) [inline src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyStringsAttribute.cs:20]: Would we want to check for IEnumerable<string?>? Even if it's nullable maybe none are.

**Review by PrincessMadMath — APPROVED** [review]

File: src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyStringsAttribute.cs (hunk 2 above)
Comment (nicholas-renaud) [reply src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyStringsAttribute.cs:20]: That would be unreachable, it equates string with string? and binds to the first one in the switch, the compiler won't even allow it

Comment (meziantou) [reply src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyStringsAttribute.cs:20]: To be safer, for the compiler, you could use

```suggestion
        IEnumerable<string?> enumerable => enumerable.All(x => !string.IsNullOrWhiteSpace(x)),
```

File: src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyGuidsAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    public override bool IsValid(object? value) => value switch
+    {
+        null => true,
+        IEnumerable<Guid> enumerable => enumerable.All(this.IsValidGuid),
+        IEnumerable<Guid?> enumerable => enumerable.All(x => x != null && this.IsValidGuid(x.Value)),
+        IEnumerable<string> enumerable => enumerable.All(this.IsValidGuid),
+        _ => false,
+    };
+
+    private bool IsValidGuid(string valueAsString)
```
Comment (meziantou) [inline src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyGuidsAttribute.cs:26]:
```suggestion
    private bool IsValidGuid(string? valueAsString)
```

File: src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyGuidsAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    public ContainsOnlyNonEmptyGuidsAttribute() : base(ErrorMessageFormat)
+    {
+    }
+
+    public override bool IsValid(object? value) => value switch
+    {
+        null => true,
+        IEnumerable<Guid> enumerable => enumerable.All(this.IsValidGuid),
+        IEnumerable<Guid?> enumerable => enumerable.All(x => x != null && this.IsValidGuid(x.Value)),
+        IEnumerable<string> enumerable => enumerable.All(this.IsValidGuid),
```
Comment (meziantou) [inline src/Workleap.ComponentModel.DataAnnotations/ContainsOnlyNonEmptyGuidsAttribute.cs:22]:
```suggestion
        IEnumerable<string?> enumerable => enumerable.All(this.IsValidGuid),
```

**Review by meziantou — APPROVED** [review]

**Pull request merged**